### PR TITLE
[Feature] Search Usecase

### PR DESCRIPTION
**Pull request by OrbitalVitamin**
This PR adds the search usecase functionality.

[Modifying firebase fetching in all repositories](https://github.com/CSC207-2022F-UofT/course-project-fitbook/commit/8b7eb288bfb64e4ba602172dfcfcf2b2ec798915)

[Add functionality for search option](https://github.com/CSC207-2022F-UofT/course-project-fitbook/commit/778d91add8dbbf555c4ff091b706694bbe6b18f9)

[Add search unit tests](https://github.com/CSC207-2022F-UofT/course-project-fitbook/commit/5de76ddb04aef9b5b969fc5fbe23bf44786a7217)

[Add search controller](https://github.com/CSC207-2022F-UofT/course-project-fitbook/commit/00fc16557e11b86b462ec511732d7cf00e463d2e)

[Add search option ports](https://github.com/CSC207-2022F-UofT/course-project-fitbook/commit/1522f0dccc2a727eb5026bc97dac8cdd5ab5fc9f)

[Add SearchResponse class](https://github.com/CSC207-2022F-UofT/course-project-fitbook/commit/bc43d73e0328a63f149c0d2a0ed12485e48e0cf5)

[add repository modifications for search usecase](https://github.com/CSC207-2022F-UofT/course-project-fitbook/commit/81c9047917851faa295dbf63249497639866d779)

[Add SearchController java](https://github.com/CSC207-2022F-UofT/course-project-fitbook/commit/d89748fcd3dff3f46c01687f4dbc71d913ce51a4)

[Delete SearchPostResponse class](https://github.com/CSC207-2022F-UofT/course-project-fitbook/commit/72da467189f59248bfd08d4adf1b875c83281e52)

[Add search functionality in SearchService](https://github.com/CSC207-2022F-UofT/course-project-fitbook/commit/37e9f9edacfbf38956020f6f36dd8fda9bf65dc1)

[Fixed Search usecase tests](https://github.com/CSC207-2022F-UofT/course-project-fitbook/commit/fde3007305a53df753c471a482bf324ac1e11f3b)

[Added documentation to code](https://github.com/CSC207-2022F-UofT/course-project-fitbook/commit/acd570812015656ac0b30f6b56552b85f8d7aca0)

### CODE REVIEW

**Review by modaserMoj — APPROVED** [review]
Great! The use case is fully implemented following all the guidelines

**Review by Ludosss — DISMISSED** [review]
Passes tests, looks good to me

File: src/main/resources/templates/search.html [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+            body: JSON.stringify({
+                followeeId: followeeId
+            })
+        }).then(() => {
+            followers += 1;
+            followStatus.textContent = "Following"
+            followButton.disabled = true;
+            followerCount.textContent = followers;
+        }).catch(e => console.error(e))
+    }
```
Comment (youssefsoli) [inline src/main/resources/templates/search.html:74]: This method is not used and can be removed

**Review by youssefsoli — CHANGES_REQUESTED** [review]

**Review by feiwun — APPROVED** [review]
code follows clean architecture and is well written

**Review by youssefsoli — APPROVED** [review]
Great implementation of complex use case! Effectively followed clean architecture and ran extensive tests covering a whole variety of different scenarios, as well as nicely documented code! Well done!

**Review by Ludosss — APPROVED** [review]
Looks great, functionality and front-end are nice. Good job!

**Pull request merged**